sm_td.h

```diff
@@ -52,679 +61,1128 @@
 
-                return false;
-            }
-            return true;
+        SMTD_DEBUG_OFFSET_INC;
+        switch (next_state->action_required) {
+            case SMTD_ACTION_TOUCH:
+                smtd_handle_action(next_state, SMTD_ACTION_TOUCH);
+                break;
+            case SMTD_ACTION_TAP:
+                smtd_handle_action(next_state, SMTD_ACTION_TOUCH);
+                smtd_handle_action(next_state, SMTD_ACTION_TAP);
+                break;
+            case SMTD_ACTION_HOLD:
+                smtd_handle_action(next_state, SMTD_ACTION_TOUCH);
+                smtd_handle_action(next_state, SMTD_ACTION_HOLD);
+                break;
+            case SMTD_ACTION_RELEASE:
+                smtd_handle_action(next_state, SMTD_ACTION_TOUCH);
+                smtd_handle_action(next_state, SMTD_ACTION_HOLD);
+                smtd_handle_action(next_state, SMTD_ACTION_RELEASE);
+                break;
+        }
+        SMTD_DEBUG_OFFSET_DEC;
 
-        case SMTD_STAGE_RELEASE:
-            // At this stage we have just released the macro key and still holding the following key
+        SMTD_DEBUG("%s %s is complete",
+                   smtd_state_to_str(state),
+                   smtd_action_to_str(action));
+    }
+}
 
-            if (keycode == state->macro_keycode && record->event.pressed) {
-                DO_ACTION_TAP(state);
+void smtd_execute_action(smtd_state *state, smtd_action action) {
+    if (state->desired_keycode == 0) {
+        state->desired_keycode = smtd_current_keycode(&state->pressed_keyposition);
+    }
 
-                SMTD_SIMULTANEOUS_PRESSES_DELAY
-                smtd_press_following_key(state, false);
+    SMTD_DEBUG("%s exec in progress with %s",
+               smtd_state_to_str(state),
+               smtd_action_to_str(action));
 
-                //todo need to go to NONE stage and from NONE jump to TOUCH stage
-                SMTD_SIMULTANEOUS_PRESSES_DELAY
-                smtd_next_stage(state, SMTD_STAGE_TOUCH);
+#ifdef SMTD_GLOBAL_MODS_PROPAGATION_ENABLED
+    uint8_t mods_on_start = get_mods();
 
-                state->sequence_len = 0;
+    if (state->saved_mods != mods_on_start) {
+        set_mods(state->saved_mods);
+        send_keyboard_report();
+        SMTD_SIMULTANEOUS_PRESSES_DELAY
+    }
 
-                return false;
-            }
-            if (
-                    keycode != state->macro_keycode
-                    && (state->following_key.row == record->event.key.row &&
-                        state->following_key.col == record->event.key.col)
-                    && !record->event.pressed
-                    ) {
-                // Following key is released. Now we definitely know that macro key is held
-                // we need to execute hold the macro key and execute tap the following key
-                // then close the state
+    uint8_t mods_on_restore = state->saved_mods;
+#else
+    uint8_t mods_on_start = get_mods();
```

Comment:
These two variables (`mods_on_start` and `mods_on_restore`) are unused, and qmk-vial compilation fails with:

```
sm_td.h:1023:13: error: unused variable 'mods_on_restore' [-Werror=unused-variable]
 1023 |     uint8_t mods_on_restore = mods_on_start;
      |             ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
 [ERRORS]
```